### PR TITLE
perf: segregate fungible assets from non-fungible assets

### LINKED ISSUE #6289: Performance improvement on the asset selection modal
### Overview

As part of the account selection ticket, we are going to also modify the way the asset selection modal works.

The functionality will work like this:
1. The chains will be in a dropdown, with 1 chain selected by default (ETH in this case)
2. Below the search and chain drop down, we show quick shortcuts for top 6 assets. This will be a pill version of the asset, so we show icon, and symbol. And possibly show their recently selected assets here??
3. In the list, we will show the users assets grouped together with a heading.
4. Below user's assets we will show the top 10-100 assets.

### References and additional details

Asset search modal

### Acceptance Criteria

- Asset search defaults to one chain at a time
- Users can easily and quickly select top assets from the quick selection menu
- Users can see all of their assets grouped together
- We don't try and preload all 10,000 assets we support

### Need By Date

_No response_

### Screenshots/Mockups

https://www.figma.com/file/IBz3Fn75GGjXfm89FsymAL/Limit-Orders?type=design&node-id=989-7670&mode=design&t=huScTA3Tv6trI9fS-0

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Improves app boot performance by segregating our static list of fungible assets from the dynamic list of non-fungible assets and discovered "unsupported" fungible assets in the user portfolio:
* Massively reduces the compute required to select the assets like because the static list is not invalidated every time an asset is added to the list. The dynamic list is very small so selector reselect is cheap.
* Reduces time to load since the static list of fungible assets is no longer tied to any network requests

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #6289

Very high risk. Affects everything in the app since this is a rework of how we manage some of the most fundamental data in the application.

> What protocols, transaction types or contract interactions might be affected by this PR?

Every page and component in the app which displays numbers, assets, balances etc.

## Testing

This PR should introduce no functional change from production, side from changes to asset sorting in select places.

* When user holds a fungible asset not in our static list of assets, it's still included in their portfolio (engineers, please test this with a monkey patch)
* Fungble asset lists are correct (e.g trade page, lp, earn, lending pages, portfolio etc)
* Non-fungible asset lists are correct (e.g portfolio, ntfs etc)
* Asset search includes all assets

### Engineering

Please assist ops in patching app locally to test difficult edge cases

### CODE REVIEW

**Comment by woodenfurniture** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/shapeshift/web/6281?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#6281** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6281?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6272** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6230** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6176** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6176?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6213** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6171** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6171?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/6281?utm_source=stack-comment).

**Comment by woodenfurniture** [comment]
Marking as draft until LP ships to mitigate risk during rollout

File: src/components/Graph/RainbowChart/RainbowChart.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ import { useSelector } from 'react-redux'
 import { Amount } from 'components/Amount/Amount'
 import { AssetIcon } from 'components/AssetIcon'
 import type { RainbowData } from 'hooks/useBalanceChartData/useBalanceChartData'
-import { selectAssets, selectSelectedLocale } from 'state/slices/selectors'
+import { selectFungibleAssets, selectSelectedLocale } from 'state/slices/selectors'
```
Comment (gomesalexandre) [inline src/components/Graph/RainbowChart/RainbowChart.tsx:16]: Feels like non-fungibles *should* be accounted for here?

File: src/context/AppProvider/AppContext.tsx [hunk 2]
```diff
@@ -65,12 +62,6 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
   // track anonymous portfolio
   useMixpanelPortfolioTracking()
 
-  // immediately load all assets, before the wallet is even connected,
-  // so the app is functional and ready
-  // if we already have assets in store, we don't need to refetch the base assets, as these won't change
-  // if they do, it means we regenerated generatedAssetData.json, and can run a migration to trigger a refetch of base assets
-  useGetAssetsQuery(undefined, { skip: Boolean(assetIds.length) })
-
```
Comment (gomesalexandre) [inline src/context/AppProvider/AppContext.tsx:73]: 👀

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 3]
```diff
@@ -331,7 +331,7 @@ type UseBalanceChartData = (args: UseBalanceChartDataArgs) => UseBalanceChartDat
 */
 export const useBalanceChartData: UseBalanceChartData = args => {
   const { assetId, accountId, timeframe } = args
-  const assets = useAppSelector(selectAssets)
+  const assets = useAppSelector(selectFungibleAssets)
```
Comment (gomesalexandre) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:334]: Even though we currently don't have market data per se for non-fungible, should we assume we can get *some* kind market data for non-fungible? Which we do as floor price in the NFT tab, even though we don't upsert it as market data:

image

See Zerion for instance, which do exclude it from the net worth, but still have a tab for NFTs with total floor price value

image

File: src/hooks/useBalanceChartData/utils.ts [hunk 4]
```diff
@@ -17,7 +17,7 @@ export const makeBalanceChartData = (total: HistoryData[] = []): BalanceChartDat
 })
 
 export const useUnavailableBalanceChartDataAssetNames = (assetIds: AssetId[]): string => {
-  const assets = useSelector(selectAssets)
+  const assets = useSelector(selectFungibleAssets)
```
Comment (gomesalexandre) [inline src/hooks/useBalanceChartData/utils.ts:20]: Same question as above here

File: src/hooks/useTxDetails/useTxDetails.ts [hunk 5]
```diff
@@ -106,7 +106,7 @@ export const getTransfers = (
 
 export const useTxDetails = (txId: string): TxDetails => {
   const tx = useAppSelector((state: ReduxState) => selectTxById(state, txId))
-  const assets = useAppSelector(selectAssets)
+  const assets = useAppSelector(selectFungibleAssets)
```
Comment (gomesalexandre) [inline src/hooks/useTxDetails/useTxDetails.ts:109]: We should be able to `useTxDetails` on NFT Txs

File: src/pages/Assets/Assets.tsx [hunk 6]
```diff
@@ -1,19 +1,17 @@
-import { useMemo } from 'react'
 import { useTranslate } from 'react-polyglot'
 import { AssetSearch } from 'components/AssetSearch/AssetSearch'
 import { Main } from 'components/Layout/Main'
 import { SEO } from 'components/Layout/Seo'
-import { selectAssetsSortedByMarketCapUserCurrencyBalanceAndName } from 'state/slices/common-selectors'
+import { selectFungibleAssetsSortedByName } from 'state/slices/selectors'
```
Comment (gomesalexandre) [inline src/pages/Assets/Assets.tsx:5]: NFT assets are assets and should be visible in the assets page

File: src/pages/TransactionHistory/DownloadButton.tsx [hunk 7]
```diff
@@ -64,6 +70,10 @@ export const DownloadButton = ({ txIds }: { txIds: TxId[] }) => {
     [translate],
   )
 
+  const assets = useMemo(() => {
+    return Object.assign({}, fungibleAssets, nonFungibleAssets)
+  }, [fungibleAssets, nonFungibleAssets])
+
```
Comment (gomesalexandre) [inline src/pages/TransactionHistory/DownloadButton.tsx:76]: Feels like this should be its own selector?

File: src/pages/TransactionHistory/hooks/useSearch.tsx [hunk 8]
```diff
@@ -3,11 +3,11 @@ import debounce from 'lodash/debounce'
 import { useCallback, useMemo, useState } from 'react'
 import { filterAssetsBySearchTerm } from 'components/TradeAssetSearch/helpers/filterAssetsBySearchTerm/filterAssetsBySearchTerm'
 import { isSome } from 'lib/utils'
-import { selectAssets } from 'state/slices/selectors'
+import { selectFungibleAssets } from 'state/slices/selectors'
 import { useAppSelector } from 'state/store'
 
 export const useSearch = () => {
-  const assetsById = useAppSelector(selectAssets)
+  const assetsById = useAppSelector(selectFungibleAssets)
```
Comment (gomesalexandre) [inline src/pages/TransactionHistory/hooks/useSearch.tsx:10]: NFT Txs are Txs and we should be able to search them like any other Tx

File: src/state/apis/zapper/zapperApi.ts [hunk 9]
```diff
@@ -150,7 +150,7 @@ export const zapperApi = createApi({
         const { data: res } = await axios.request({ ...payload })
         const zapperV2AppTokensData = V2AppTokensResponse.parse(res)
 
-        const assets = selectAssets(getState() as ReduxState)
+        const assets = selectNonFungibleAssets(getState() as ReduxState)
```
Comment (gomesalexandre) [inline src/state/apis/zapper/zapperApi.ts:153]: This is wrong - this query fetches UNI-V2 pool assetIds, which are fungible

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 10]
```diff
@@ -105,17 +123,18 @@ export const assets = createSlice({
...
     clear: () => initialState,
-    upsertAssets: (state, action: PayloadAction<UpsertAssetsPayload>) => {
-      state.byId = Object.assign({}, state.byId, action.payload.byId) // upsert
-      state.ids = Array.from(new Set(state.ids.concat(action.payload.ids)))
+    upsertNonFungibleAssets: (state, action: PayloadAction<UpsertAssetsPayload>) => {
+      state.nonFungible.byId = Object.assign({}, state.nonFungible.byId, action.payload.byId) // upsert
+      state.nonFungible.ids = Object.keys(state.nonFungible.byId)
     },
-    upsertAsset: (state, action: PayloadAction<Asset>) => {
+    upsertNonFungibleAsset: (state, action: PayloadAction<Asset>) => {
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/assetsSlice.ts:130]: We should still be able to fetch fungibles - which we do when e.g upserting UNI-V2 assets data, or asset description on this same file on l.170:

https://github.com/shapeshift/web/blob/25c6ece1c1c628f4392594095a8f9413d8eb0a10/src/state/slices/assetsSlice/assetsSlice.ts#L170

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 11]
```diff
@@ -162,17 +151,23 @@ export const assetApi = createApi({
...
+          const data = {
+            byId: {
+              [assetId]: originalAsset
+                ? Object.assign(originalAsset, { description, isTrusted })
+                : undefined,
+            },
+            ids: [assetId],
+          }
+
+          if (data) dispatch(assets.actions.upsertNonFungibleAssets(data))
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/assetsSlice.ts:170]: This is fungible asset data

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 12]
```diff
@@ -124,36 +143,6 @@ export const assetApi = createApi({
   ...BASE_RTK_CREATE_API_CONFIG,
   reducerPath: 'assetApi',
   endpoints: build => ({
-    getAssets: build.query<UpsertAssetsPayload, void>({
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/assetsSlice.ts:127]: Feels like we shouldn't remove this?

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 13]
```diff
@@ -277,7 +277,7 @@ describe('portfolioSlice', () => {
 
         // dispatch asset data
         const assetData = mockAssetState()
-        store.dispatch(assetsSlice.actions.upsertAssets(assetData))
+        store.dispatch(assetsSlice.actions.upsertNonFungibleAssets(assetData))
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:280]: This is fungible assets data

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 14]
```diff
@@ -329,7 +329,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:332]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 15]
```diff
@@ -387,7 +387,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:390]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 16]
```diff
@@ -467,7 +467,7 @@ describe('portfolioSlice', () => {
 
       // dispatch asset data
       const assetData = mockAssetState()
-      store.dispatch(assetsSlice.actions.upsertAssets(assetData))
+      store.dispatch(assetsSlice.actions.upsertNonFungibleAssets(assetData))
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:470]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 17]
```diff
@@ -526,7 +526,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:529]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 18]
```diff
@@ -595,7 +595,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:598]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 19]
```diff
@@ -669,7 +669,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:672]: Ditto

File: src/state/slices/portfolioSlice/portfolioSlice.test.ts [hunk 20]
```diff
@@ -716,7 +716,7 @@ describe('portfolioSlice', () => {
...
```
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSlice.test.ts:719]: Ditto

File: src/state/slices/search-selectors.ts [hunk 21]
```diff
@@ -34,7 +34,7 @@ export type GlobalSearchResult = AssetSearchResult | TxSearchResult | SendResult
 export type SelectGlobalItemsFromFilterReturn = [AssetSearchResult[], TxSearchResult[]]
 
 export const selectGlobalItemsFromFilter = createDeepEqualOutputSelector(
-  selectAssetsBySearchQuery,
+  selectFungibleAssetsBySearchQuery,
```
Comment (gomesalexandre) [inline src/state/slices/search-selectors.ts:37]: Ditto NFT assets are assets and should be searchable

File: src/state/slices/txHistorySlice/selectors.ts [hunk 22]
```diff
@@ -219,7 +219,7 @@ export const selectMaybeNextAccountNumberByChainId = createSelector(
 export const selectTxsByQuery = createDeepEqualOutputSelector(
   selectTxs,
   selectTxIdsByFilter,
-  selectAssets,
+  selectFungibleAssets,
```
Comment (gomesalexandre) [inline src/state/slices/txHistorySlice/selectors.ts:222]: Ditto Txs with NFTs are Txs and should be searchable

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
@woodenfurniture see testing below - many many regressions with this guy and probably a lot more than I've found out here.

While I conceptually get the intent re: perf here, I don't think this is feasible in a non-breaking fashion? i.e:

> Massively reduces the compute required to select the assets like because the static list is not invalidated every time an asset is added to the list. The dynamic list is very small so selector reselect is cheap.

The issue here is that a recompute *is* expected. As we upsert asset description and upsert assets meta (e.g Zapper UNI-V2 opportunities meta), assets *should* recompute so they are reflected at view-layer. Same goes for NFTs, and failure to do so will result in missing assets meta, and consequently user data.

Given how extensive this diff is because of the LSP rename and the number of places where fungible/non-fungible was forgotten, if we want to keep going with this PR, I would suggest to at the very least:

- keep `selectAssets` as a selector encapsulating fungible and non-fungible, and *don't* update consumption ensuring we still consume the same assets it in most places, except in ones where this is a huge perf. issue (e.g asset selection) 
- ensure there are no regressions in upsertion, see e.g comments in `<AppContext />` and RTK queries

However at this stage, most of the benefits would arguably be gone, with a high possiblity of regressions still present, so happy to discuss this PR further.

---

Tested locally (after a local merge of `origin/develop` which I didn't do in this PR to avoid restacking issues, and after a full cache clear):

- NFT assets are now missing in wallet in this diff

image
image

- Idle and UNI-V2 assets are now missing in this diff

image

- Searching for NFTs is now broken

image

- Manually navigating to an NFT asset page it's now obvious the meta isn't fetched for it, most likely the cause of all the issues above and most likely related to the removal of https://github.com/shapeshift/web/pull/6281/files#r1537409080

 
image

Note missing balance entry, and missing symbol in this screenshot

- Asset descriptions is now missing (tested with FOXy, ATOM, FOX and ETH)

image
image
image
image

**Comment by woodenfurniture** [comment]
Following converstation with @gomesalexandre , the decision is to close this in favour of new smaller tasks which should achieve the same / better gain with less risk:
Move uni v2 pools into generate-assets script #6559
Add staticAssets as a new field in assets slice #6560
Sort the static generated assets list by display name #6561

**Comment by woodenfurniture** [comment]
Following converstation with @gomesalexandre , the decision is to close this in favour of new smaller tasks which should achieve the same / better gain with less risk:
Move uni v2 pools into generate-assets script https://github.com/shapeshift/web/issues/6559
Add staticAssets as a new field in assets slice https://github.com/shapeshift/web/issues/6560
Sort the static generated assets list by display name https://github.com/shapeshift/web/issues/6561
